Implement DeepLC

RalfG opened the pull request:
- New `config_parser` module with support for TOML files and legacy MS²PIP txt config files
- Implemented DeepLC predictions in new `retention_time` module, also implemented in fasta2speclib
- Small fasta2speclib improvements

RalfG committed:
Add pytestcache to gitignore

RalfG committed:
Move other HCD models to archive folder

RalfG committed:
Initial tests with new ConfigParser module and toml format

RalfG committed:
Fully implement new ConfigParser

RalfG committed:
Merge branch 'releases' into feature/implement_deeplc

RalfG committed:
Bugfix after merge

RalfG committed:
Add tomlkit requirement

RalfG committed:
Add retention_time module

RalfG committed:
Implement RT preds in ms2pip and fasta2speclib

RalfG committed:
Update readme

RalfG committed:
Bugfix

RalfG committed:
Lazy-load DeepLC

RalfG committed:
Use more recent build action and upload artifacts

paretje commented on README.md:
```diff
@@ -99,6 +99,8 @@ optional arguments:
   -s MGF_FILE           .mgf MS2 spectrum file (optional)
   -w FEATURE_VECTOR_OUTPUT
                         write feature vectors to FILE.{pkl,h5} (optional)
+  --retention_time      add retention time predictions (requires DeepLC python
```
Usually dashes are used instead of underscores in argument names

RalfG committed:
Add long CLI options

RalfG committed:
Fix tests

paretje commented on config.toml:
```diff
@@ -0,0 +1,29 @@
+verbose=false
+path_model=false # Path to DeepLC model
+split_cal=25 # Can maximally be number of calibration peptides - 1
+batch_num=350000 # Reduce to limit memory consumption
```
We probably should have a consistent formatting here, i.e. be consistent about `x=y` vs `x = y`.

RalfG replied:
The TOML specification examples seem to use whitespaces: https://github.com/toml-lang/toml, so we'll use that

paretje commented on fasta2speclib/fasta2speclib.py:
```diff
@@ -86,35 +87,52 @@ def get_params():
+            is_valid = False
+        else:
+            is_valid = True
+        return is_valid
```
Just assign the expression to the variable without if/else, or you could even just return it.

paretje commented on fasta2speclib/fasta2speclib.py:
```diff
@@ -381,23 +407,24 @@ def run_batches(peprec, decoy=False):
+        #    complevel=3,
+        #    complib="zlib",
+        #    mode="a",
+        #)
```
Remove commented code?

paretje commented on ms2pip/models/HCD/hcd_fast_B.c:
```diff
@@ -1,3046 +0,0 @@
-static float score_HCD_B(unsigned int* v){
```
`ms2pip/cython_modules/ms2pipfeatures_c_HCDch2.c` contains a reference (in comments) to these deleted files, that should be removed as well.

paretje commented on ms2pip/ms2pipC.py:
```diff
@@ -883,8 +873,11 @@ def __init__(
+                self.config_parser.load()
+                self.params = self.config_parser.config['ms2pip']
         else:
+            self.config_parser = None
```
Another option could be to always pass the config as a params dict to the runner, and move the config parser logic to main.

RalfG replied:
Indeed. Removed `config_file` argument from MS2PIP class, and moved all config parser logic to `__main__.py`. One consequence of this (required for deeplc integration, and fasta2speclib in the future) is that params is now a dict with `{"ms2pip": <ms2pip_params>}` or `{"ms2pip": <ms2pip_params>, "deeplc": <deeplc_params>}`

paretje commented on ms2pip/ms2pipC.py:
```diff
@@ -946,10 +939,18 @@ def run(self):
+                rt_predictor = RetentionTime(config=self.config_parser.config)
+            else:
+                rt_predictor = RetentionTime()
+            rt_predictor.add_rt_predictions(self.data)
```
Not something that should be done in this PR, but one option to improve the runner class and method would be to split the different logical actions to different methods, i.e. instead of having a `match_spectrum` option, or deferring actions from passed MGF files.

paretje reviewed (changes requested):
Some unused models have been moved to the archive folder. Why not just remove them and rely on git as an archive?

RalfG committed:
Merge branch 'releases' into feature/implement_deeplc

RalfG committed:
Changes according to PR review
- fasta2speclib validate_peptide reversed logic
- remove unused cython modules
- Move config parsing to __main__.py
- Refactor test_features and test_predictions according to MS2PIP module and new config_parsing

RalfG committed:
Fix tests

paretje commented on ms2pip/ms2pipC.py:
```diff
@@ -683,48 +662,44 @@ def __init__(
         output_filename=None,
         datasetname=None,
-        return_results=True,
+        return_results=False,
```
Maybe we should drop `return_results`, and introduce `write_results`, and always return the results. This should probably done when we refactor `MS2PIP.run`.

RalfG committed:
Remove old models, small fixes
